src/transformers/modeling_reformer.py

```diff
@@ -0,0 +1,1757 @@
+
+        if isinstance(config.hidden_act, str):
+            self.act_fn = ACT2FN[config.hidden_act]
+        else:
+            self.act_fn = config.hidden_act
+
+        self.dense = nn.Linear(config.hidden_size, config.feed_forward_size)
+
+    def forward(self, hidden_states):
+        hidden_states = self.dense(hidden_states)
+        hidden_states = nn.functional.dropout(hidden_states, self.dropout, self.training)
+        hidden_states = self.act_fn(hidden_states)
+        return hidden_states
+
+
+class ReformerFeedForwardOutput(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.dropout = config.hidden_dropout_prob
+
+        self.dense = nn.Linear(config.feed_forward_size, config.hidden_size)
+
+    def forward(self, hidden_states):
+        hidden_states = self.dense(hidden_states)
+        hidden_states = nn.functional.dropout(hidden_states, self.dropout, self.training)
+        return hidden_states
+
+
+class ChunkReformerFeedForward(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.chunk_size_feed_forward = config.chunk_size_feed_forward
+        self.seq_len_dim = 1
+
+        self.layer_norm = ReformerLayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+        self.dense = ReformerFeedForwardDense(config)
+        self.output = ReformerFeedForwardOutput(config)
+
+    def forward(self, attention_output):
+        return apply_chunking_to_forward(
+            self.chunk_size_feed_forward, self.seq_len_dim, self.forward_chunk, attention_output,
+        )
+
+    def forward_chunk(self, hidden_states):
+        hidden_states = self.layer_norm(hidden_states)
+        hidden_states = self.dense(hidden_states)
+        return self.output(hidden_states)
+
+
+class ReformerLayer(nn.Module):
+    def __init__(self, config, layer_id=0):
+        super().__init__()
+        self.attention = ReformerAttention(config, layer_id)
+        # dropout requires to have the same
+        # seed for forward and backward pass
+        self.attention_seed = None
+        self.feed_forward_seed = None
+
+        self.feed_forward = ChunkReformerFeedForward(config)
+
+    def _init_attention_seed(self):
+        """
+            This function sets a new seed for the
+            attention layer to make dropout deterministic
+            for both forward calls: 1 normal forward
+            call and 1 forward call in backward
+            to recalculate activations.
+        """
+
+        # randomize seeds
+        if next(self.parameters()).device.type == "cuda":
+            # GPU
+            device_idx = torch.cuda.current_device()
+            self.attention_seed = torch.cuda.default_generators[device_idx].seed()
+            torch.cuda.manual_seed(self.attention_seed)
+        else:
+            # CPU
+            self.attention_seed = int(torch.seed() % sys.maxsize)
+            torch.manual_seed(self.attention_seed)
+
+    def _init_feed_forward_seed(self):
```

Comment:
this code seems nearly identical to prev function if you move the attribute setting logic to `__init__`